### ISSUE TITLE
TF dataset: support for aspect ratio group batching

### ISSUE THREAD

**Issue by MrWanter** [tensorflow/tensorflow#19710]
Current TF dataset class only support fixed size padding for batching different sized input elements, however there are cases like in object detection, we may need to batch images according to their aspect ratios, i.e., push images with similar aspect ratio into one batch and pad them with minimal extra paddings, is there any existing approach that does not use feed_dict approach but directly utilize TF operators and read TF record files for accomplish this ? Thanks!

**Comment by mrry** [comment]
This should be possible using [`tf.contrib.data.group_by_window()`](https://www.tensorflow.org/api_docs/python/tf/contrib/data/group_by_window) to group the images into batches based on their aspect ratios, and then apply the appropriate padding for the aspect ratio in the window function. (This is the same mechanism that’s used to [bucket sequences by length](https://www.tensorflow.org/api_docs/python/tf/contrib/data/bucket_by_sequence_length), and looking at the [implementation](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/data/python/ops/grouping.py#L95) of that feature might be useful.)

**Comment by MrWanter** [comment]
Thank you @mrry Could you provide a simple example of using this function, I mean how to set 

```
(key_func,
 reduce_func,
 window_size=None,
 window_size_func=None)
```
for my case ?

**Comment by MrWanter** [comment]
specifically, to write key_func, we have to calculate the aspect ratio of each elem, but the shape of image elem is (?,?,3) as image width and height are varying, in this case, how to write the key_func ? I have searched many examples, but they based on fixed small amount of input data, for which statistics like length, width and height are readily available, but for tf-record files the decoded data shape usually has none type, do I have to reconvert the dataset to add an aspect ratio feature for each image?

**Comment by MrWanter** [comment]
batch_size_training = 4
bucket_boundaries_ = [0.3, 0.5, 0.7, 0.9, 1.0, 1.1, 1.3, 1.5, 1.7]
bucket_batch_sizes_ = (len(bucket_boundaries_)+1)*[batch_size_training]

def element_length_function(elem):
    height, width = tf.cast(tf.shape(elem['image'])[0], tf.float32), \
                    tf.cast(tf.shape(elem['image'])[1], tf.float32)
    ratio = width/height
    return ratio
def bucket_by_sequence_length(element_length_func = element_length_function,
                              bucket_boundaries= bucket_boundaries_,
                              bucket_batch_sizes = bucket_batch_sizes_,
                              padded_shapes=None,
                              padding_values=None,
                              pad_to_bucket_boundary=False):

    with ops.name_scope("bucket_by_seq_length"):
        if len(bucket_batch_sizes) != (len(bucket_boundaries) + 1):
            raise ValueError(
                "len(bucket_batch_sizes) must equal len(bucket_boundaries) + 1")

        batch_sizes = constant_op.constant(bucket_batch_sizes, dtype=dtypes.int64)

        def element_to_bucket_id(*args):
            """Return int64 id of the length bucket for this element."""
            seq_length = element_length_func(*args)

            boundaries = list(bucket_boundaries)
            buckets_min = [np.iinfo(np.int32).min] + boundaries
            buckets_max = boundaries + [np.iinfo(np.int32).max]
            conditions_c = math_ops.logical_and(
                math_ops.less_equal(buckets_min, seq_length),
                math_ops.less(seq_length, buckets_max))
            bucket_id = math_ops.reduce_min(array_ops.where(conditions_c))

            return bucket_id



        def window_size_fn(bucket_id):
            # The window size is set to the batch size for this bucket
            window_size = batch_sizes[bucket_id]
            return window_size


        def make_padded_shapes(shapes, none_filler=None):
            padded = []
            for shape in nest.flatten(shapes):
                shape = tensor_shape.TensorShape(shape)
                shape = [
                    none_filler if d.value is None else d
                    for d in shape
                ]
                padded.append(shape)
            return nest.pack_sequence_as(shapes, padded)


        def batching_fn(bucket_id, grouped_dataset):
            """Batch elements in dataset."""
            batch_size = batch_sizes[bucket_id]
            none_filler = None
            if pad_to_bucket_boundary:
                err_msg = ("When pad_to_bucket_boundary=True, elements must have "
                           "length <= max(bucket_boundaries).")
                check = check_ops.assert_less(
                    bucket_id,
                    constant_op.constant(len(bucket_batch_sizes) - 1,
                                         dtype=dtypes.int64),
                    message=err_msg)
                with ops.control_dependencies([check]):
                    boundaries = constant_op.constant(bucket_boundaries,
                                                      dtype=dtypes.int64)
                    bucket_boundary = boundaries[bucket_id]
                    none_filler = bucket_boundary
            shapes = make_padded_shapes(
                padded_shapes or grouped_dataset.output_shapes,
                none_filler=none_filler)
            return grouped_dataset.padded_batch(batch_size, shapes, padding_values)


        def _apply_fn(dataset):
            return dataset.apply(
                tf.contrib.data.group_by_window(element_to_bucket_id, batching_fn,
                                window_size_func=window_size_fn))

        return _apply_fn

@mrry, I tried to modify the  bucket sequences by length, but it only generate one image a time, can you pls point where I'm wrong?

**Comment by MrWanter** [comment]
OK, it works!

**Comment by mrry** [comment]
Great! Glad to hear it's working. If you have other questions about how to use the API, I'd recommend posting on Stack Overflow, under the `[tensorflow-datasets]` tag.

**Comment by MrWanter** [comment]
@mrry Thank you, one more thing, since reading the lower level code could be a bit hard, could you briefly explain how group_by_window is implemented, I read pytorch implementation of faster-rcnn with aspect ratio grouping, they scan the whole dataset and group images with similar aspect ratio into batches, every epoch, shuffle is done between batches, not every image, is group_by_window doing the same thing?

**Comment by ppwwyyxx** [comment]
Jumping in with a minor comment.. @MrWanter just saw that you've been posting many dataset-related issues for the past week. This is exactly why I wrote this in tensorpack tutorial:
> Unlike running a mathematical model, data processing is a complicated and poorly-structured task. You need to handle different formats, handle corner cases, noisy data, combination of data. Doing these requires condition operations, loops, data structures, sometimes even exception handling. These operations are naturally not the right task for a symbolic graph.

It depends on your exact task but for Faster-RCNN my experience is that a Python loader can run much faster than the speed of training.

**Comment by MrWanter** [comment]
@ppwwyyxx Hi, Yuxin, I see, thanks for sharing.
